File: go/test/commons_test.go
package test

import (
	"math"
	. "shapes/base/commons"
	helper "shapes/test/fixtures"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestMeasurableInterface(test *testing.T) {
	test.Run("Test Circle Implements Measurable", func(test *testing.T) {
		circle := helper.SetupCircle()
		assertMeasurable(test, &circle, 754.767, 97.389)
	})

	test.Run("Test Rectangle Implements Measurable", func(test *testing.T) {
		rectangle := helper.SetupRectangle()
		assertMeasurable(test, &rectangle, 756.437, 112.0)
	})

	test.Run("Test Square Implements Measurable", func(test *testing.T) {
		square := helper.SetupSquare()
		assertMeasurable(test, &square, 105.062, 41.0)
	})
}

func assertMeasurable(
	test *testing.T,
	shape Dimensional,
	expectedArea float64,
	expectedPerimeter float64,
) {
	var areaDiff = math.Abs(expectedArea - shape.Area())
	var perimeterDiff = math.Abs(expectedPerimeter - shape.Perimeter())

	assert.True(test, areaDiff < DELTA, "AREA NOT EQUAL")
	assert.True(test, perimeterDiff < DELTA, "PERIMETER NOT EQUAL!")
}

File: ruby/test/commons_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require "minitest/reporters"
require "shapes/base/commons"

require_relative "fixtures/helper"

Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new(color: true)

class TestDimensional < Minitest::Test
    def test_circle_implements_dimensional
        circle = Helper.setup_circle
        assert_dimensional(circle, 754.767, 97.389)
    end

    def test_rectangle_implements_dimensional
        rectangle = Helper.setup_rectangle
        assert_dimensional(rectangle, 756.437, 112.0)
    end

    def test_square_implements_dimensional
        square = Helper.setup_square
        assert_dimensional(square, 105.062, 41.0)
    end

    private

    def assert_dimensional(
        dimensional_shape,
        expected_area,
        expected_perimeter
    )
        area_diff = (dimensional_shape.area - expected_area).abs
        perimeter_diff = (dimensional_shape.perimeter - expected_perimeter).abs

        assert(area_diff < 0.001, "AREA NOT EQUAL")
        assert(perimeter_diff < 0.001, "PERIMETER NOT EQUAL!")
    end
end
